[영화 극장 선택] 호두 3,4단계 미션 제출합니다.

Junyoung-WON opened the pull request:
안녕하세요! 두루!
3, 4단계의 예매내역 저장, 알림 기능 미션 제출합니다.
하지만 기능 구현에 먼저 힘쓰느라, 리팩터링과 테스트 코드가 제대로 되어있지 않습니다.
그리고 알람 기능도 지금 제대로 동작하지 않는 것 같아요...
시간이 없으니 먼저 PR을 보내고, 이후에도 계속 기능 수정과 리팩터링 진행하겠습니다.
우선 현재 기능 구현한 코드까지만 피드백을 해주시면 감사하겠습니다!

## 개선해야 할 점

### 알림 클릭 시 실제 예매한 시간과 다름
알림은 설정한 시간에 정상적으로 오지만, 알림을 통해 예매 확인 화면으로 이동하면, 다른 시간대가 나타납니다. 지금 데이터로서는 09:00이 상영 시작으로 고정되어 나타나고 있습니다. 아직 원인을 찾지 못했는데, 다음 PR에서 꼭 개선하겠습니다!

### 복잡한 도메인(상영 시간 및 날짜)
티켓 미션 때에 도메인 설계를 너무 깊고 복잡하게 진행하여서, 현재 불필요한 기능들도 많고, 이에 따라 위와 같은 오류들이 나타나는 것 같습니다. 도메인 설계를 조금 더 가볍게 변경할 필요가 있는 것 같습니다.

### 알림 권한 설정 기능
알림 권한을 설정하는 기능이 현재 동작하지 않고 있습니다. 앱 자체 설정 화면에서 설정하든 하지 않든 알림을 계속 받을 수 있습니다.


## 커밋 비교 링크
https://github.com/Junyoung-WON/android-movie-theater/compare/f76e985034dad4ac7f22698b5ba14077406f9fe3..89298cfb9896f84a98384ab591fa068469808625

Junyoung-WON committed:
docs: 예매 내역을 확인하는 기능 요구 사항 작성 (step3)

Junyoung-WON committed:
build: Room 데이터베이스 라이브러리를 사용하기 위한 의존성 추가

Junyoung-WON committed:
feat: 극장별 영화 상영 날짜와 시간에 따라, 스피너에서 상영 시간을 선택하도록 기능 구현

Junyoung-WON committed:
feat: 예매한 티켓 정보를 Room DB에 저장할 Database 및 Dao, Entity 객체 설정

RoomDatabase 설정
- Entity: Reservation
- Dao: ReservaionDao
- Database: ReservationDatabase

Converters 추가 구현
- List 형식의 데이터를 저장할 수 있는 TypeConverter 구현
- build.gradle.kts에 gson 의존성 추가

Junyoung-WON committed:
feat: 예매 내역 화면을 구성할 Layout xml 파일 작성

item_reservation_detail.xml
- 예매 내역 아이템을 나타냄
- data binding 활용
- 관련 dimens 및 strings 리소스 추가

fragment_reservation_details.xml
- 파일 이름 변경
- RecyclerView 태그 추가

Junyoung-WON committed:
feat: 예매 내역 Fragment 구현

ReservationDetailsContract
- 예매 내역 화면의 View와 Presenter 사이의 인터페이스를 정의한 Contract 작성

ReservationDetailsListAdapter, DetailViewHolder
- 예매 내역 RecyclerView의 어댑터와 뷰홀더 구현
- 데이터 바인딩 활용

ReservationDetailsFragment, ReservationDetailsPresenterImpl
- Presenter와 Fragment(View)의 인터페이스 구현

MovieTicketUiModel
- DB의 Entity로부터 UiModel로 변환하는 팩토리 함수 추가

NavigationActivity
- Fragment 클래스 이름 변경에 따른 리팩터링

Junyoung-WON committed:
feat: 좌석 선택 완료 시 Room DB에 예매 정보를 저장하는 기능 추가

Junyoung-WON committed:
refactor: Kotlin의 thread 블럭 활용 및 데이터 Load 동작의 순서 수정

Junyoung-WON committed:
docs: 상영 푸시 알림 설정 기능 요구사항 작성 (step 4)

Junyoung-WON committed:
feat: 알림 채널 및 builder 설정하여 알림을 시작하는 MovieStartNotificationManager

TODO: ID, 제목, 설명 등 문자열 값을 리소스를 활용한 리팩터링 필요

Junyoung-WON committed:
feat: 알람 매니저가 시스템에 전달할 액션(알림)을 수신하는 AlarmReceiver 생성

TODO: 액션(알림)에 대한 문자열 리소스화 필요

Junyoung-WON committed:
feat: 영화 시작 30분 전 시스템에 알람을 전달하는 AlarmManager 작성

TODO: Intent의 Extra Key 상수화 및 알람 해제 기능 작성 필요

Junyoung-WON committed:
feat: 예매 완료 시 AlarmManager와 AlarmReceiver 활용하여 알림 설정

Junyoung-WON committed:
feat: 알람 확인을 위한 상영 날짜 데이터 추가

Junyoung-WON committed:
build: 설정 기능을 위한 preference 의존성 추가

Junyoung-WON committed:
feat: 알람 설정 화면 구현

Junyoung-WON committed:
refactor: 테스트를 위한 영화 상영 날짜 범위 수정

Junyoung-WON committed:
feat: 앱 시작 시 알림 권한에 대한 기능 구현

Junyoung-WON committed:
refactor: 표준 시간대를 시스템과 동일하게 수정 및 Intent에 action 추가

Junyoung-WON committed:
refactor: Intent의 키 값 상수 사용

Junyoung-WON committed:
refactor: EXAXT_ALARM에 대한 권한 추가

Junyoung-WON committed:
fix: 날짜 변경 시 시간 대가 자동으로 선택되지 않는 오류 수정

Junyoung-WON committed:
fix: 예매 완료 시에만 알람을 설정할 수 있도록 intent extra 추가 및 리소스 활용

Junyoung-WON committed:
style: Log 제거 및 import 순서 수정

Junyoung-WON committed:
fix: Room에 저장되는 티켓 ID는 자동 증가이므로 id값을 넘겨주지 않아도 됨

Junyoung-WON committed:
refactor: 불필요한 xml 제거

Gyuil-Hwnag commented on README.md:
```diff
@@ -12,3 +12,24 @@
     - 홈 : 영화 리스트
     - 설정 : 빈 화면
+
+## STEP 3
```
README 작성 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/AndroidManifest.xml:
```diff
@@ -1,6 +1,10 @@
-    xmlns:tools="http://schemas.android.com/tools">
+    xmlns:tools="http://schemas.android.com/tools" >
+
+    <uses-permission android:name="android.permission.POST_NOTIFICATIONS" />
+    <uses-permission android:name="android.permission.SCHEDULE_EXACT_ALARM" />
+    <uses-permission android:name="android.permission.USE_EXACT_ALARM" />
```
권한 설정 좋습니다~👍
권한 관련 키워드가 나왔으니, 한번 알아두고 가면 좋을 내용이 있어서 남겨두었습니다!
Android는 OS별로도 권한이 사라지고, 변경이 되는 경우가 있어서 이를 위해 maxSdkVersion등의 설정도 추가할 수 있습니다!
https://developer.android.com/guide/topics/manifest/uses-permission-element

Junyoung-WON replied:
좋은 레퍼런스 공유 감사드립니다!! 😄

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/alarm/AlarmReceiver.kt:
```diff
@@ -0,0 +1,22 @@
+import android.content.Intent
+import woowacourse.movie.R
+
+class AlarmReceiver : BroadcastReceiver() {
+    override fun onReceive(
```
AlarmReceiver관련 작성 잘해주셨네요~👍
추후 시간이 남을 때 BroadcastReceiver() 관련해서 한번 살펴보고 가는 것도 좋을 것 같다는 생각이 듭니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/alarm/MovieStartAlarmManager.kt:
```diff
@@ -0,0 +1,62 @@
+        val alarmTime = dateTimeBefore30min()
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            alarmManager.setExactAndAllowWhileIdle(
+                AlarmManager.RTC_WAKEUP,
+                alarmTime,
+                pendingIntent,
+            )
```
PR에도 남겨주셨듯이, 알림이 정확한 시간에 안오는 이슈가 있더라구요🥲
현재 단계에서는 수정이 필요한 내용은 아니며, 추후 호두가 알림 관련해서 개발을 하실 때 도움이 될 만한 내용인 것 같아서 코멘트 남겨두었습니다!💪

`alarmManager.setExactAndAllowWhileIdle()`의 경우 정확한 시간에 알림이 안울리는 이슈가 있습니다.
시간을 정확하게 하기 위해서`alarmManager.canScheduleExactAlarms()`를 사용한다면, 사용자가 앱에 대해서 `알림 및 리마인더`를 설정해줘야 합니다.
`사용자에게 알림 허용 -> 알림 및 리마인더 설정 2댑스` vs `시간이 부정확하더라도 1댑스로 기능 활성화` 와 같은 고민이 생기실텐데, 어떤 구현을 하는것이 좋을지 한번 고민해보면 추후에 도움이 될 것 같다는 생각이 듭니다!

이전에 저도 고민 했던 내용이 있어서 링크 첨부하였습니다!
[알림 기능1](https://github.com/Gyuil-Hwnag/compose_study/pull/24)
[알림 기능2](https://github.com/Gyuil-Hwnag/compose_study/pull/25)

Junyoung-WON replied:
AlarmManager에 리시버에게 보낼 알람을 설정하는 메서드가 정말 다양하게 있었네요...! 정확한 시간에 울리는 알람은 사용자가 리마인더에 대한 설정을 해주어야 하는군요. 꼭 참고하여서, 나중에 관련 기능을 구현할 때 잘 고민하고 활용해보겠습니다😊

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/alarm/MovieStartNotificationManager.kt:
```diff
@@ -0,0 +1,67 @@
+            notificationBuilder = NotificationCompat.Builder(context, channelId)
+        } else {
+            notificationBuilder = NotificationCompat.Builder(context)
+        }
```
버전별 분기처리 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/alarm/MovieStartNotificationManager.kt:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        notificationBuilder.setSmallIcon(R.drawable.ic_home)
+            .setContentTitle("영화")
+            .setContentText("영화 상영 시작 30분 전입니다.")
```
해당 하드코딩으로 들어간 내용에 대해서는 Android Resource를 활용하도록 해보는 것은 어떤가 라는 생각이 듭니다!💪

Junyoung-WON replied:
기능에 신경쓰다가 리팩터링을 못하고 제출했네요😅 string 리소스를 활용하도록 해보겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/HomeFragment.kt:
```diff
@@ -31,18 +31,19 @@ class HomeFragment : Fragment(), ScreeningContract.View, ScreeningContract.ViewA
         super.onViewCreated(view, savedInstanceState)
-        val movieListView = view.findViewById<RecyclerView>(R.id.movieList)
-        movieListView.layoutManager = LinearLayoutManager(context)
-        movieListView.adapter = adapter
+        binding.movieList.layoutManager = LinearLayoutManager(context)
```
`binding.movieList.layoutManager = LinearLayoutManager(context)` 해당 내용은 xml 내에서도 선언이 가능하니 참고 부탁드립니다~💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/NavigationActivity.kt:
```diff
@@ -10,13 +16,23 @@ class NavigationActivity : AppCompatActivity() {
         findViewById(R.id.navigationView)
     }
 
+    private val requestPermissionLauncher =
+        registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted: Boolean ->
+            if (isGranted) {
+                showToastMessage(MESSAGE_PUSH_STATE_ON)
+            } else {
+                showToastMessage(MESSAGE_PUSH_STATE_OFF)
+            }
+        }
+
```
Permission 에 대한 처리 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/NavigationActivity.kt:
```diff
@@ -10,13 +16,23 @@ class NavigationActivity : AppCompatActivity() {
         if (savedInstanceState == null) {
             bottomNavigationView.selectedItemId = R.id.home_fragment
         }
+        requestNotificationPermission()
     }
```
Notification 권한에 대한 여부를 앱이 실행될 때 묻도록 설정을 하신 이유가 궁금합니다🤔
사용성 관점인 것 같아서 호두의 자유로운 의견을 공유해주세요~💪

Junyoung-WON replied:
영화가 시작되기 30분 전임을 알려주는 알림은 구현한 앱이 갖고 있는 편의 기능 중 하나입니다.
일부 사용자는 이런 기능을 필요로 하고, 해당 기능이 정상적으로 잘 동작하기를 원할 수도 있다고 생각합니다.
그래서 알림 기능의 동작 여부를 확실히 정할 수 있도록, 앱을 실행했을 때에 사용자에게 알림에 대한 권한을 묻는 것이 더 적합하다고 생각했습니다..!

Junyoung-WON replied:
물론 이러한 구현은 다른 일부 사용자들의 사용 경험을 떨어뜨릴 수 있을 것 같습니다.
그래서 예약이 완료되고 난 후 예매 결과를 확인하는 화면에서 알림에 대한 권한을 얻도록 묻는 것도 좋은 방법이 될 것 같네요!

Gyuil-Hwnag replied:
@Junyoung-WON 
타당한 의견 좋습니다~
아마 실무에서는 이러한 권한을 묻는 정책들에 대해서, 고민을 해볼 기회가 있을 것 같은데 그럴 때 호두의 생각을 잘 얘기해본다면 좋은 사용자 경험을 제공할 수 있을 것 같습니다😀

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/NavigationActivity.kt:
```diff
@@ -50,4 +66,31 @@ class NavigationActivity : AppCompatActivity() {
+                Manifest.permission.POST_NOTIFICATIONS,
+            ) != PackageManager.PERMISSION_GRANTED
+
+        if (isPermissionNotGranted) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+                if (shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS)) {
+                    showToastMessage(MESSAGE_PUSH_STATE_OFF)
+                } else {
+                    requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
+                }
```
버전별 처리 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/SettingFragment.kt:
```diff
@@ -1,18 +1,45 @@
+        val isNotificationEnabled = sharedPreferences?.getBoolean("notifications", false)
+        if (rootKey != null) {
+            notificationPreference = findPreference("notifications")
+        }
```
Activity 에서 Preference에 대해서 직접적으로 접근을 하지 않고, 함수또는 객체를 통해서 설정을 해보도록 하는 것은 어떤가 라는 생각이 듭니다!💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/SettingFragment.kt:
```diff
@@ -1,18 +1,45 @@
-    override fun onCreateView(
-        inflater: LayoutInflater,
-        container: ViewGroup?,
+class SettingFragment : PreferenceFragmentCompat() {
+    private var notificationPreference: SwitchPreferenceCompat? = null
```
`PreferenceFragmentCompat()`의 활용 너무 잘해주셨습니다~
`PreferenceFragmentCompat()`를 활용하지 않고 구현을 하려면 어떻게 해야 할지 한번 고민해보면 좋을 것 같다는 생각이 듭니다!💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/detail/MovieDetailPresenterImpl.kt:
```diff
@@ -18,6 +18,9 @@ class MovieDetailPresenterImpl(
     val movieId: Int,
     val theaterId: Int,
```
movieId, theaterId 를 Preference를 인스턴스 생성시에 부 생성자로 반드시 받아야 하는 이유가 있을지 궁금합니다!🤔

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/result/ReservationResultActivity.kt:
```diff
@@ -1,29 +1,52 @@
 
     private fun setUpFromIntent() {
-        val ticket = intent.getParcelableExtra<MovieTicketUiModel>(INTENT_TICKET)
+        ticket = intent.getParcelableExtra<MovieTicketUiModel>(INTENT_TICKET)!!
         binding.data = ticket
```
non-nullable 하다고 강제롤 하는 !!는 자칫 오류를 유발하는 구현이 될 수 있겠다는 생각이 듭니다🥲
해당 내용을 좀 더 유연하게 처리해 볼 수 있을 것 같다는 생각이 듭니다!
추가적으로 `getParcelableExtra()`에 대한 버전별 분기처리도 해보면 좋을 것 같다는 생각이 듭니다~💪

Gyuil-Hwnag commented on app/src/main/res/values/dimens.xml:
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
+    <dimen name="textsize_item_movietitle">16dp</dimen>
+    <dimen name="padding_item_reservation_detail">12dp</dimen>
+    <dimen name="margin_item_span">5dp</dimen>
+    <dimen name="margin_item_between_title_date">6dp</dimen>
```
dimen의 활용 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/seat/SeatSelectionPresenterImpl.kt:
```diff
@@ -2,13 +2,17 @@ package woowacourse.movie.presentation.view.reservation.seat
 import woowacourse.movie.presentation.uimodel.MovieTicketUiModel
+import kotlin.concurrent.thread
 
 class SeatSelectionPresenterImpl(
     reservationCount: Int,
```
Presenter에 대한 전반적인 코멘트 입니다~
Presenter에 대한 테스트도 작성을 해보면 좋을 것 같다는 생각이 듭니다!💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -23,6 +24,7 @@ import woowacourse.movie.presentation.view.reservation.detail.MovieDetailActivit
 import woowacourse.movie.presentation.view.reservation.result.ReservationResultActivity
 
 class SeatSelectionActivity : BaseActivity(), SeatSelectionContract.View, SeatSelectionContract.ViewActions {
+    private lateinit var reservationDatabase: ReservationDatabase
```
새롭게 추가된 Activity, Fragment에 대한 전반적인 코멘트 입니다!
Activity, Fragment에 대한 테스트도 작성을 해보면 좋을 것 같다는 생각이 듭니다!💪

Gyuil-Hwnag reviewed (changes requested):
호두 안녕하세요!

해당 미션 구현 고생 많으셨습니다!
이전 1, 2 단계의 리뷰 반영하면서 3, 4의 내용 구현 너무 잘해주셨습니다 💯

말씀해주신대로, 현재 시간이 부족하여 3, 4단계 구현에 대해서 치중을 하였으니 이번 요청에서는 리팩토링, 리뷰 반영, 테스트 코드를 중심으로 살펴보면 좋을 것 같다는 생각이 듭니다!
고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

추가적으로 모든 내용을 반영하다 보면, 시간이 모자랄 수 있으니 호두의 판단하에 중요하다고 생각되는 우선순위를 정하여 반영을 해보는 것도 좋을 것 같습니다😀
(추후 반영할 내용들에 대해서는 고민후에 코멘트를 남겨두고 차차 반영을 해나가는 방식도 좋을 것 같다는 생각이 듭니다)

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 
화이팅입니다💪

Junyoung-WON committed:
refactor: RecyclerView의 LayoutManager 설정을 xml에 작성

Junyoung-WON committed:
refactor: SharedPreferences를 싱글톤으로 활용

Junyoung-WON committed:
refactor: 캐싱된 sharedPreference를 가져오는 로직을 변경

Junyoung-WON committed:
refactor: sharedPreference로부터 알림 설정 여부를 가져와 알림 발송

Junyoung-WON committed:
refactor: strings 리소스 활용

Junyoung-WON committed:
fix: 과거의 알림이 계속해서 나타나는 오류 수정

pending Intent 설정에서의 requestCode 값과 Flag 값 변경
- requestCode 값을 랜덤하게 설정하여, 이전의 알림과 중복되지 않도록 변경
- Flag에 UPDATE_CURRENT 값을 설정하여 이전의 알림을 가져오지 않도록 변경

Junyoung-WON committed:
refactor: IntentCompat 활용하여 버전별 intent 사용 분기 처리

Junyoung-WON committed:
refactor: pendingIntent를 설정하는 메서드명 수정 및 null값에 대한 처리

Junyoung-WON committed:
build: ui test 및 presenter 테스트를 위한 테스트 라이브러리 의존성 추가

Junyoung-WON committed:
test: presenter 테스트 작성

Junyoung-WON committed:
test: ui 테스트 작성

Junyoung-WON commented:
안녕하세요, 두루!
극장 미션 피드백 제출이 너무 늦어졌네요.
이전에 구현 못했던 기능 요구사항을 최대한 만족하고
UI테스트와 Presenter 테스트도 전부는 아니지만 작성해보았습니다.
지금 진행중인 미션에도 집중하고싶어서
이전 미션은 여기까지만 진행하고 마무리짓고 싶습니다...!
두루와 많은 피드백을 주고받지 못해서 너무 아쉽지만,
그럼에도 두루가 제게 가르쳐주신 것들 모두 감사하게 생각하고 있습니다!
두루가 알려주신 몇가지 팁들과 조언들을 활용해서
이후 미션에서도 잘 적용할 수 있도록 열심히 공부하겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/SettingFragment.kt:
```diff
@@ -1,18 +1,37 @@
-        inflater: LayoutInflater,
-        container: ViewGroup?,
+class SettingFragment : PreferenceFragmentCompat() {
+    private val sharedPref by lazy { MovieSharedPreferences(requireContext()) }
+    private var notificationPreference: SwitchPreferenceCompat? = null
```
`Context`라는 키워드가 나왔으니, Context 라는 키워드에 대해서 한번 정리를 해본다면, 추후에 많은 도움이 될 것 같다는 생각이 듭니다~💪
추가적으로 requireContext()는 어떤 Context를 가져오는지에 대해서 살펴본다면, 해당 인스턴스가 싱글톤으로 사용되는지 좀 더 명확하게 구현을 해볼 수 있겠다는 생각이 듭니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/sharedpreference/MovieSharedPreferences.kt:
```diff
@@ -0,0 +1,28 @@
+import android.content.Context
+import android.content.SharedPreferences
+
+class MovieSharedPreferences(context: Context) {
+    private val prefs = getSharedPreferences(context)
```
호두가 SharedPreference를 사용하려고 했던 의도에 맞게 사용을 하려면, class 가 나을지, object 가 나을지도 한번 살펴보면 좋을 것 같다는 생각이 듭니다~💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/alarm/MovieStartNotificationManager.kt:
```diff
@@ -0,0 +1,74 @@
+        )
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val channelId = context.getString(R.string.notification_channel_id)
+            val name = context.getString(R.string.notification_channel_name)
+            val descriptionText = context.getString(R.string.notification_channel_description)
```
Android Resource의 활용 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/alarm/MovieStartNotificationManager.kt:
```diff
@@ -0,0 +1,74 @@
+            MovieTicketUiModel::class.java,
+        )
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val channelId = context.getString(R.string.notification_channel_id)
```
버전별 분기처리 잘해주셨네요~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/alarm/MovieStartNotificationManager.kt:
```diff
@@ -0,0 +1,74 @@
+            .setPriority(NotificationCompat.PRIORITY_MAX)
+            .setAutoCancel(true)
+
+        if (ticket != null) {
+            setPendingIntent(notificationBuilder, ticket)
+        }
+        ticket?.ticketId?.let { notificationManager.notify(it.toInt(), notificationBuilder.build()) }
```
해당 내용은 모두 ticket이 null 이 나일 때 사용되는 것이므로, 하나로 같이 묶어 보는 형식으로 한다면, 하단에 있는 불필요한 null check 를 없애볼 수 있겠다는 생각도 듭니다~💪

Gyuil-Hwnag commented on app/src/test/java/woowacourse/movie/presenter/MovieDetailPresenterTest.kt:
```diff
@@ -0,0 +1,104 @@
+import woowacourse.movie.presentation.view.reservation.detail.MovieDetailPresenterImpl
+import woowacourse.movie.repository.TheaterRepositoryImpl
+
+class MovieDetailPresenterTest {
+    private lateinit var view: MovieDetailContract.View
```
Presenter에 대한 테스트 코드 잘 작성해주셨네요~👍

Gyuil-Hwnag commented on app/src/test/java/woowacourse/movie/presenter/MovieDetailPresenterTest.kt:
```diff
@@ -0,0 +1,104 @@
+        verify { view.updateScreeningTimes(any(), any()) }
+    }
+
+    @Test
+    fun `예매 인원 감소 버튼을 누르면 view에게 변경된 예매 수를 보여줄 것을 요청한다`() {
+        // given
+        presenter.plusReservationCount()
+
+        // when
+        presenter.minusReservationCount()
+
+        // then
+        verify { view.updateReservationCount(1) }
+    }
```
given, when, then 으로 테스트 함수를 너무 잘 구현해주셨네요~👍

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/presentation/view/reservation/seat/SeatSelectionActivityTest.kt:
```diff
@@ -0,0 +1,198 @@
+import woowacourse.movie.presentation.view.seatSelectionActivityIntent
+import java.util.regex.Pattern
+
+class SeatSelectionActivityTest {
+    private val testContext = ApplicationProvider.getApplicationContext<Context>()
+    private val intent: Intent = seatSelectionActivityIntent(testContext)
```
추후에 시간이 난다면, Fragment 에 대한 테스트 코드를 작성해봐도 좋은 경험이 될 것 같다는 생각이 듭니다~👍

Gyuil-Hwnag reviewed (approved):
호두 안녕하세요!
미션 너무 고생 많으셨습니다! 👏

리뷰 반영 정말 잘 해주셨어요💯
구현을 하시면서 여러가지 고민을 하셨을 텐데 다음 미션 에서도 잘 적용을 하면서 발전해 나가시길 바라겠습니다😆

다음 미션 진행을 위하여 해당 미션은 여기서 머지하도록 하겠습니다.
도움이 될만한 내용들 코멘트 달았으니 확인해보셔도 좋을 것 같습니다!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
앞으로도 응원하겠습니다 💪
